File: solutions/_547_number_of_provinces/solution.ts
/**
 * @param {number[][]} isConnected
 * @return {number}
 */
function findCircleNum(isConnected: number[][]) {
  let numCities = isConnected.length;
  const visited = new Set<number>();
  let numProvinces = 0;

  for (let i = 0; i < numCities; i++) {
    if (!visited.has(i)) {
      dfs(isConnected, visited, i);
      numProvinces++;
    }
  }

  return numProvinces;
}

function dfs(isConnected: number[][], visited: Set<number>, i: number) {
  visited.add(i);
  for (let j = 0; j < isConnected.length; j++) {
    if (isConnected[i][j] === 1 && !visited.has(j)) {
      dfs(isConnected, visited, j);
    }
  }
}

export const solution = findCircleNum;

/*
https://leetcode.com/problems/number-of-provinces/
*/

File: solutions/_547_number_of_provinces/go/solution.go
package _547_number_of_provinces

func findCircleNum(isConnected [][]int) int {
	n := len(isConnected)
	visited := make([]bool, n)
	count := 0

	for i := 0; i < n; i++ {
		if !visited[i] {
			dfs(isConnected, visited, i)
			count++
		}
	}

	return count
}

func dfs(isConnected [][]int, visited []bool, i int) {
	visited[i] = true
	for j := 0; j < len(isConnected); j++ {
		if isConnected[i][j] == 1 && !visited[j] {
			dfs(isConnected, visited, j)
		}
	}
}

/*
https://leetcode.com/problems/number-of-provinces/
*/
